Issue with periodic deletion of banner pictures

```
May 16 11:00:27 obey java[22055]: Caused by: org.postgresql.util.PSQLException: ERROR: update or delete on table "banner_picture" violates foreign key constraint "banner_picture_fkey" on table "banner"
May 16 11:00:27 obey java[22055]:   Detail: Key (hash)=(3d119a946c0cc0a32c6f6d1ef259106d) is still referenced from table "banner".
May 16 11:00:27 obey java[22055]:         at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2553) ~[postgresql-42.2.19.jar!/:42.2.19]
May 16 11:00:27 obey java[22055]:         at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2285) ~[postgresql-42.2.19.jar!/:42.2.19]
May 16 11:00:27 obey java[22055]:         at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:323) ~[postgresql-42.2.19.jar!/:42.2.19]
May 16 11:00:27 obey java[22055]:         at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:481) ~[postgresql-42.2.19.jar!/:42.2.19]
May 16 11:00:27 obey java[22055]:         at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:401) ~[postgresql-42.2.19.jar!/:42.2.19]
May 16 11:00:27 obey java[22055]:         at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:164) ~[postgresql-42.2.19.jar!/:42.2.19]
May 16 11:00:27 obey java[22055]:         at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:130) ~[postgresql-42.2.19.jar!/:42.2.19]
May 16 11:00:27 obey java[22055]:         at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61) ~[HikariCP-3.4.5.jar!/:na]
May 16 11:00:27 obey java[22055]:         at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java) ~[HikariCP-3.4.5.jar!/:na]
May 16 11:00:27 obey java[22055]:         at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197) ~[hibernate-core-5.4.29.Final.jar!/:5.4.29.Final]
May 16 11:00:27 obey java[22055]:         ... 34 common frames omitted
```

--- a/src/main/java/com/bannergress/backend/entities/Banner.java
+++ b/src/main/java/com/bannergress/backend/entities/Banner.java
@@ -114,7 +114,7 @@ public class Banner {
     /**
      * Generated Picture.
      */
-    @OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
     @JoinColumn(name = "picture")
     @NotAudited
     private BannerPicture picture;

--- a/src/main/java/com/bannergress/backend/entities/BannerPicture.java
+++ b/src/main/java/com/bannergress/backend/entities/BannerPicture.java
@@ -6,6 +6,7 @@ import net.karneim.pojobuilder.GeneratePojoBuilder;
 import javax.persistence.*;
 
 import java.time.Instant;
+import java.util.List;
 
 /**
  * Represents a banner picture.
@@ -29,6 +30,9 @@ public class BannerPicture {
     @Column(name = "expiration", nullable = true)
     private Instant expiration;
 
+    @OneToMany(mappedBy = "picture", fetch = FetchType.LAZY)
+    private List<Banner> banners;
+
     public Instant getExpiration() {
         return expiration;
     }
@@ -51,5 +55,13 @@ public class BannerPicture {
 
     public void setPicture(byte[] picture) {
         this.picture = picture;
+    }
+    
+    public List<Banner> getBanners() {
+        return banners;
+    }
+    
+    public void setBanners(List<Banner> banners) {
+        this.banners = banners;
     }
 }

--- a/src/main/java/com/bannergress/backend/services/impl/BannerPictureServiceImpl.java
+++ b/src/main/java/com/bannergress/backend/services/impl/BannerPictureServiceImpl.java
@@ -22,6 +22,7 @@ import javax.imageio.ImageWriteParam;
 import javax.imageio.ImageWriter;
 import javax.imageio.stream.ImageOutputStream;
 import javax.persistence.EntityManager;
+import javax.persistence.TypedQuery;
 
 import java.awt.AlphaComposite;
 import java.awt.Color;
@@ -32,6 +33,7 @@ import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.time.Instant;
+import java.util.List;
 import java.util.Map.Entry;
 import java.util.Optional;
 import java.util.concurrent.ExecutionException;
@@ -183,7 +185,16 @@ public class BannerPictureServiceImpl implements BannerPictureService {
 
     @Override
     public void removeExpired() {
-        entityManager.createQuery("DELETE FROM BannerPicture WHERE expiration < :now")
-            .setParameter("now", Instant.now()).executeUpdate();
+        TypedQuery<BannerPicture> query = entityManager
+            .createQuery("SELECT p FROM BannerPicture p WHERE p.expiration < :now", BannerPicture.class);
+        query.setParameter("now", Instant.now());
+        List<BannerPicture> pictures = query.getResultList();
+        for (BannerPicture picture : pictures) {
+            if (picture.getBanners().isEmpty()) {
+                entityManager.remove(picture);
+            } else {
+                picture.setExpiration(null);
+            }
+        }
     }
 }

